### PR TITLE
Add OptionalPath extractor

### LINKED ISSUE #1884: OptionalPath extractor
- [x] I have looked for existing issues (including closed) about this

### Motivation

We've been getting many people asking about an easy way of extracting a path segment value only if a placeholder exists at the route the handler was called from.

### Proposal

Add an extractor `struct OptionalPath<T>(Option<T>);` that succeeds with `None` if there are no placeholders for the given route, and if there are placeholders, attempts to deserialize the arguments to `T` and return `Some(T)` on success, rejecting the request on failure.

### Alternatives

🤷🏼

### PR DESCRIPTION
**Pull request by jplatte**
## Motivation

Resolves #1884.

## Solution

What was already written on the issue 😄 

I also had to add `kind()` to `FailedToDeserializePathParams`. I don't think we should really have both `kind()` and `into_kind()` but I wasn't too sure what it should look like and went for roughly the path of least resistance (it would have been easier to return `&ErrorKind` but I'm guessing we don't want to guarantee that the rejection holds an `ErrorKind`).

### CODE REVIEW

File: axum/src/extract/path/mod.rs [hunk 1]
```diff
@@ -380,7 +380,14 @@ impl fmt::Display for ErrorKind {
 pub struct FailedToDeserializePathParams(PathDeserializationError);
 
 impl FailedToDeserializePathParams {
+    /// Get the underlying error kind.
+    pub fn kind(&self) -> ErrorKind {
+        self.0.kind.clone()
+    }
```
Comment (davidpdrsn) [inline axum/src/extract/path/mod.rs:386]: Any reason to not return a reference here? I guess it means we change the internals of `FailedToDeserializePathParams` but maybe that's okay

```suggestion
    pub fn kind(&self) -> &ErrorKind {
        &self.0.kind
    }
```

Comment (jplatte) [reply axum/src/extract/path/mod.rs:386]: If we commit to `FailedToDeserializePathParams` containing `ErrorKind`, do we keep `into_kind` or deprecate / later remove it? If we keep the `Clone` impl I added for making the current implementation of this method possible, `.kind().clone()` would be semantically equivalent to the current `.into_kind()` (though theoretically less efficient).

Comment (jplatte) [reply axum/src/extract/path/mod.rs:386]: Ping @davidpdrsn

Comment (davidpdrsn) [reply axum/src/extract/path/mod.rs:386]: I'm fine with committing to `ErrorKind` for now.

Comment (jplatte) [reply axum/src/extract/path/mod.rs:386]: Err, I don't see how that answers my question.. I guess I will just do the easiest thing, keep `into_kind`, remove the `Clone` impl and not deprecate anything.

Comment (davidpdrsn) [reply axum/src/extract/path/mod.rs]: Ah sorry. I thought you were asking about committing to having an `ErrorKind` inside `FailedToDeserializePathParams`, not whether to keep `into_kind` and `kind`. I'm fine with keeping both methods, like you've implemented it here.

File: axum-extra/src/extract/optional_path.rs [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+///     // Convert to u32, default to page 1 if not specified
+///     let page = page.map_or(1, |param| param.get());
+///     // ...
+/// }
+///
+/// let app = Router::new()
+///     .route("/blog", get(render_blog))
+///     .route("/blog/:page", get(render_blog));
+/// # let app: Router = app;
+/// # app.into_make_service();
```
Comment (davidpdrsn) [inline axum-extra/src/extract/optional_path.rs:34]: If the previous like works then so should this. Can probably just omit it.

```suggestion
```

**Review by davidpdrsn — COMMENTED** [review]
Changelog?

**Comment by davidpdrsn** [comment]
I'll do the last view changes now.

**Review by davidpdrsn — APPROVED** [review]

**Pull request merged**